src/libcore/ptr.rs

```diff
@@ -228,6 +228,67 @@ pub const fn null<T>() -> *const T { 0 as *const T }
 #[rustc_promotable]
 pub const fn null_mut<T>() -> *mut T { 0 as *mut T }
 
+#[repr(C)]
+union Repr<T> {
+    rust: *const [T],
+    rust_mut: *mut [T],
+    raw: FatPtr<T>,
+}
+
+#[repr(C)]
+struct FatPtr<T> {
+    data: *const T,
+    len: usize,
+}
```

Comment:
Can this be shared with the `Repr` in `core::slice`? Or maybe remove the latter and implement the reference-version in terms of the raw-ptr version?